javascripts/discourse/initializers/upload-consent-initializer.js

```diff
@@ -1,38 +1,36 @@
 import { withPluginApi } from "discourse/lib/plugin-api";
-import { inject as service } from "@ember/service";
 import UploadConsent from "../components/modal/upload-consent";
 
 const uploadRegexp = /\(upload?:\/\/[\w\d./?=#]+\)/;
+import { Promise } from "rsvp";
 
 function initialize(api) {
-  api.modifyClass("service:composer", {
-    pluginId: "discourse-upload-consent",
-
-    modal: service(),
-
-    save() {
-      const enabledCategories = settings.consent_enabled_categories
-        .split("|")
-        .map((id) => parseInt(id, 10));
-
-      if (
-        enabledCategories.includes(this.model.categoryId) &&
-        uploadRegexp.test(this.model.reply)
-      ) {
-        const originalSave = this._super;
-        this.modal.show(UploadConsent, {
+  const composerService = api.container.lookup("service:composer");
+  const modal = api.container.lookup("service:modal");
+
+  const enabledCategories = settings.consent_enabled_categories
+    .split("|")
+    .map((id) => parseInt(id, 10));
+
+  api.composerBeforeSave(() => {
+    const categoryId = composerService?.model?.categoryId;
+    const reply = composerService?.model?.reply;
+
+    if (!categoryId || !reply) {
+      return Promise.resolve();
+    }
+
+    return new Promise((resolve) => {
+      if (enabledCategories.includes(categoryId) && uploadRegexp.test(reply)) {
+        modal.show(UploadConsent, {
```

Comment:
Also, do you know if the upload consent modal should be closable? i.e. clicking "cancel" shows the error message, but you can currently still click "x" to dismiss. Is that expected?

If not, then this line should be deleted: https://github.com/discourse/discourse-upload-consent/blob/5af965c021ea38626f28d5478791755809c4ed1e/javascripts/discourse/components/modal/upload-consent.hbs#L2C30-L2C30

…and then you could just `await` `modal.show()` (and the savePost arg could be removed completely)

Reply:
I guess it should be closable, in case someone disagrees with the policy.

Reply:
But clicking "Cancel" doesn't close it, so I'm not sure if that's the intent there